blockdag/reachability.go

```diff
@@ -0,0 +1,501 @@
+	queue := []*reachabilityTreeNode{rtn}
+	for len(queue) > 0 {
+		var current *reachabilityTreeNode
+		current, queue = queue[0], queue[1:]
+		if len(current.children) == 0 {
+			// We reached a leaf
+			current.subtreeSize = 1
+		}
+		if current.subtreeSize <= uint64(len(current.children)) {
+			// We haven't yet calculated the subtree size of
+			// the current node. Add all its children to the
+			// queue
+			for _, child := range current.children {
+				queue = append(queue, child)
+			}
+			continue
+		}
+
+		// We reached a leaf or a pre-calculated subtree.
+		// Push information up
+		for current != rtn {
+			current = current.parent
+			current.subtreeSize++
+			if current.subtreeSize != uint64(len(current.children)) {
+				// Not all subtrees of the current node are ready
+				break
+			}
+			// All subtrees of current have reported readiness.
+			// Count actual subtree size and continue pushing up.
+			childSubtreeSizeSum := uint64(0)
+			for _, child := range current.children {
+				childSubtreeSizeSum += child.subtreeSize
+			}
+			current.subtreeSize = childSubtreeSizeSum + 1
+		}
+	}
+	return rtn.subtreeSize
+}
+
+// applyIntervalDown applies new intervals using a BFS traversal.
+// The intervals are allocated according to subtree sizes and the
+// 'split' allocation rule (see the split() method for further
+// details)
+func (rtn *reachabilityTreeNode) applyIntervalDown(interval *reachabilityInterval) error {
+	err := rtn.setTreeInterval(interval)
+	if err != nil {
+		return err
+	}
+
+	queue := []*reachabilityTreeNode{rtn}
+	for len(queue) > 0 {
+		var current *reachabilityTreeNode
+		current, queue = queue[0], queue[1:]
+		if len(current.children) > 0 {
+			sizes := make([]uint64, len(current.children))
+			for i, child := range current.children {
+				sizes[i] = child.subtreeSize
+			}
+			intervals, err := current.remainingInterval.split(sizes)
+			if err != nil {
+				return err
+			}
+			for i, child := range current.children {
+				childInterval := intervals[i]
+				err := child.setTreeInterval(childInterval)
+				if err != nil {
+					return err
+				}
+				queue = append(queue, child)
+			}
+
+			// Empty up remaining interval
+			current.remainingInterval.start = current.remainingInterval.end + 1
+		}
+
+		// Cleanup temp info for future reindexing
+		current.subtreeSize = 0
+	}
+	return nil
+}
+
+// String returns a string representation of a reachability tree node
+// and its children.
+func (rtn *reachabilityTreeNode) String() string {
+	nodeString := rtn.interval.String()
+	queue := []*reachabilityTreeNode{rtn}
+	for len(queue) > 0 {
+		var current *reachabilityTreeNode
+		current, queue = queue[0], queue[1:]
+		if len(current.children) == 0 {
+			break
+		}
+
+		nodeString += "\n"
+		for _, child := range current.children {
+			nodeString += child.interval.String()
+			queue = append(queue, child)
+		}
+	}
+	return nodeString
+}
+
+// reachabilityInterval represents an interval to be used within the
```

Comment:
Place reachabilityInterval on top of the file, since everything depends on it, but it depends on nothing.